initial draft of rootfs rpi4 zarhus crosscon hv instructions (#17)
zarhus/docs: docs/supported-targets/rpi4-crosscon-hypervisor.md

# Raspberry Pi 4 for the CROSSCON Hypervisor

This guide provides info on how to boot a working linux system
with mounted `rootfs`, `systemd` services working, and a login
console that lets the user execute commands on a Raspberry Pi 4
device running the CROSSCON Hypervisor.

## Build CROSSCON Hypervisor

We recommend first following the
[default demo](https://github.com/crosscon/CROSSCON-Hypervisor-and-TEE-Isolation-Demos/tree/master/rpi4-ws/)
using the

> [docker environment](https://github.com/crosscon/CROSSCON-Hypervisor-and-TEE-Isolation-Demos/pull/1)

Author: @DaniilKl i added this in [this commit](https://github.com/zarhus/docs/pull/17/commits/c0bac8b93ea1ab5ae1d0ed246325e187b8eb8633)

to isolate the dependencies.

Make sure to copy over the contents of the SD card somewhere, so that
they are not lost - they will be needed later.

## Build Zarhus

> Obviously, Zarhus will also have to be built (on the host).

Reviewer: Do not use `Obviously`, try to be as professional as possible.

Author: @DaniilKl fixed [here](https://github.com/zarhus/docs/pull/17/commits/8e318e800bd588a6c134f742c6f1b3db4bf24a1e)

Make sure to run this exact build command:

```bash
SHELL=/bin/bash KAS_MACHINE=raspberrypi4-64 kas-container build meta-zarhus/kas/common.yml:meta-zarhus/kas/cache.yml:meta-zarhus/kas/debug.yml:meta-zarhus/kas/rpi.yml
```

> The `KAS_MACHINE` variable is very important here, `raspberrypi4` would build a 32-bit

Reviewer: You could use `!!! warning`, [example](https://github.com/zarhus/docs/blob/4013485640192e13a7c028fd063511aae6282a3d/docs/getting-started/flashing.md?plain=1#L44).

Author: @DaniilKl I dont think this is a good idea, this is not a warning but more of an explanation.
Also, it would ruin the formatting of the existing `code blocks`.
If you strongly feel this should be included, then ping me again in this thread and i will add this

Reviewer: Then `!!! note`, because `variable is very important here` sound like you want all attention from reader.

> version of the system - we need to use `raspberrypi4-64` in order to get a 64-bit

Reviewer: Use impersonal form.

Author: @DaniilKl fixed [here](https://github.com/zarhus/docs/pull/17/commits/3b254378887d4d24c0df1bd3b66b3a9490f55667)

version, so that it's compatible with the hypervisor.

> ## Combine the two

Reviewer: Maybe:

```suggestion
## The final image
```

?

> When the Zarhus build finishes, flash the SD card with it like this:

Reviewer: ```suggestion
When the Zarhus build finishes, flash the SD card:
```

```bash
cd build/tmp/deploy/images/raspberrypi4-64/
sudo bmaptool copy --bmap zarhus-base-image-debug-raspberrypi4-64.rootfs.wic.bmap zarhus-base-image-debug-raspberrypi4-64.rootfs.wic.gz /dev/sd[X]
```

> (replace `/dev/sd[X]` with the actual card). Once that finishes, reinsert the card
> and delete everything from the `/boot` partition, and replace it with the contents
> produced by the CROSSCON Hypervisor demo step above.

Reviewer: ```suggestion
!!! note 
    Replace `/dev/sd[X]` with the actual card.
 
Once that finishes, reinsert the card and delete everything from the `/boot` partition,
and replace it with the content produced by the CROSSCON Hypervisor demo step
above.
```

> We also need to swap the kernel for the Zarhus kernel. It can be done by first copying

Reviewer: Use impersonal form.

Author: @DaniilKl fixed [here](https://github.com/zarhus/docs/pull/17/commits/d99ac528f40430153c08cbc16f3646a366fc8995)

> We also need to swap the kernel for the Zarhus kernel. It can be done by first copying

Reviewer: ```suggestion
We also need to swap the kernel for the Zarhus OS kernel. It can be done by first copying
```

Author: @DaniilKl added [here](https://github.com/zarhus/docs/pull/17/commits/249d8e90ff4be3a4bb23fac08c9e311774099c24)

> the kernel to our container (make sure to copy the actual kernel, not the symlink):

Reviewer: Use impersonal form.

Author: @DaniilKl fixed [here](https://github.com/zarhus/docs/pull/17/commits/e475624512d8bfc1ae01ff6aab59cec7c24790ec)

```bash
cd build/tmp/deploy/images/raspberrypi4-64/

> file Image # use the output of this command for the next one

Reviewer: How exactly the output should be used?

Author: @DaniilKl added [here](https://github.com/zarhus/docs/pull/17/commits/ae52686ffb71519c6809c6b16363efa5ea54e785)

docker cp Image-1-6.6.22+git0+6a24861d65_c04af98514-r0-raspberrypi4-64-20250226164115.bin crosscon_hv_container:/work/Image
```

> We will have to recompile the Hypervisor, starting from step 9 of the original demo.

Reviewer: You could link the step to make it more convenient for end user.

Author: @DaniilKl i linked this specific step from a specific commit [here](https://github.com/zarhus/docs/pull/17/commits/832b5c02fe9df494dd65ac80eba13bbdecaaff65)

> First apply this change to the device tree file in `rpi4-ws` folder:

Reviewer: You should mention that you are doing it inside a container.

Author: @DaniilKl added [here](https://github.com/zarhus/docs/pull/17/commits/c8245d74447b9d20678592ca33b68ab586238d7f)

```patch

> root@565810a48049:/work/crosscon/rpi4-ws# git --no-pager diff rpi4.dts

Reviewer: You are showing diff but not how to apply the patch.

Author: @DaniilKl
That's because I'm not really showing a patch.

I'm showing the user what the finished product needs to be like.

Remember, people build the hypervisor with different configurations all of the time.

Here I am showing what change needs to be made. I am not specifically instructing anyone to use a patch,
since the repo might differ based on what changes one has locally.

The user might just use a text editor to apply this, or if they wish they can copy this code block,
turn it into a patch and then try to apply that.

Point is, i believe this is a much better approach than just adding a patch file. If has made any changes at all to the `.dts` file (which is done very often), it would break the patch.

I think this approach is good - if you disagree with my reasoning, please provide some idea of a better way to move forward with this and I will add it.

Reviewer: Ok. You should then explain how to fix problem with `apt ` and install `vim` or smth similar, because this container does not have `vim` or `vim`.

Author: @DaniilKl added [here](https://github.com/zarhus/docs/pull/17/commits/c72ccdcf9874a8d79790c4a3ace0f114209eedba)

diff --git a/rpi4-ws/rpi4.dts b/rpi4-ws/rpi4.dts
index 0a690a0..3d06bbc 100644
--- a/rpi4-ws/rpi4.dts
+++ b/rpi4-ws/rpi4.dts
@@ -18,7 +18,7 @@

     chosen {
         stdout-path = "serial1:115200n8";
-        bootargs = "earlycon clk_ignore_unused ip=192.168.42.15 carrier_timeout=0";

> +        bootargs = "8250.nr_uarts=8 root=/dev/mmcblk1p2 rw rootwait console=ttyS1,115200 earlycon clk_ignore_unused ip=192.168.42.15 carrier_timeout=0";

Reviewer: Why this `8250.nr_uarts=8`? Please, explain the changes.

Reviewer: What `8250.nr_uarts` does mean? Explain the change in the document too.

Author: @DaniilKl changes explained [here](https://github.com/zarhus/docs/pull/17/commits/efb58f3214b46fed99dd03cf328397a583d99f15)

     };

     reserved-memory {
@@ -824,7 +824,7 @@
             pinctrl-names = "default";
             pinctrl-0 = <0x07 0x08>;
             uart-has-rtscts;
-            status = "okay";
+            status = "disabled";

             bluetooth {
                 compatible = "brcm,bcm43438-bt";
@@ -1402,14 +1402,14 @@
         #address-cells = <0x02>;
         #size-cells = <0x01>;
         ranges = <0x00 0x7e000000 0x00 0xfe000000 0x1800000>;
-        dma-ranges = <0x00 0xc0000000 0x00 0x00 0x40000000>;
+        dma-ranges = <0x0 0x0 0x0 0x0 0xfc000000>;

         emmc2@7e340000 {
             compatible = "brcm,bcm2711-emmc2";
             reg = <0x00 0x7e340000 0x100>;
             interrupts = <0x00 0x7e 0x04>;
             clocks = <0x06 0x33>;
-            status = "disabled";
+            status = "okay";
             vqmmc-supply = <0x1e>;
             vmmc-supply = <0x1f>;
             broken-cd;

> root@565810a48049:/work/crosscon/rpi4-ws#

Reviewer: This patch does not apply for me:

```bash
root@f60583377683:/work/crosscon# git apply patch
error: patch failed: rpi4-ws/rpi4.dts:18
error: rpi4-ws/rpi4.dts: patch does not apply
root@f60583377683:/work/crosscon/rpi4-ws# git apply patch
error: patch failed: rpi4-ws/rpi4.dts:18
error: rpi4-ws/rpi4.dts: patch does not apply
```

Please, specify which commits you are using for all repos and submodules inside the container.

Author: @DaniilKl well i build the container from this:
```
wgrzywacz in ~/crosscon-3mdeb/CROSSCON-Hypervisor-and-TEE-Isolation-Demos on network_support ● λ git rev-parse HEAD
61b5cae83061a0098abe0fc39f4ed25bef4429d0
wgrzywacz in ~/crosscon-3mdeb/CROSSCON-Hypervisor-and-TEE-Isolation-Demos on network_support ● λ 
```

and here's what's inside the container:
```
root@565810a48049:/work/crosscon# git remote -v
origin	git@github.com:3mdeb/CROSSCON-Hypervisor-and-TEE-Isolation-Demos.git (fetch)
origin	git@github.com:3mdeb/CROSSCON-Hypervisor-and-TEE-Isolation-Demos.git (push)
root@565810a48049:/work/crosscon# git rev-parse HEAD
7e63ce2560d2259b6af4536b2de0d133edd1b6ab
root@565810a48049:/work/crosscon# git submodule foreach --recursive 'echo $path: $(git rev-parse HEAD)'
Entering 'CROSSCON-Hypervisor'
CROSSCON-Hypervisor: e50c585d7a3c9b32bb61ca0dfdab61055e3cdaa4
Entering 'bitcoin-wallet'
bitcoin-wallet: 1e4b05ebc7538dab483dd919d3d7c06f534a51ed
Entering 'linux'
linux: 0eb9b5b5f26d459b125dd89a9b7364d5d170b53a
Entering 'opensbi'
opensbi: 4489876e933d8ba0d8bc6c64bae71e295d45faac
Entering 'optee_client'
optee_client: 932a1ceb4e7ee3bc5559f92a939687aded4307bd
Entering 'optee_os'
optee_os: c30e42806629891dc9511468c290571deb505fa8
Entering 'optee_test'
optee_test: 3cc6173b5812935adb071bd7b7ecfa42cb7cd2c5
root@565810a48049:/work/crosscon# 
```

Reviewer: Ok, I was able to reproduce your results.

```

> then run these commands:

Reviewer: ```suggestion
Then run the following commands:
```

```bash
cd /work/crosscon
dtc -I dts -O dtb rpi4-ws/rpi4.dts > rpi4-ws/rpi4.dtb
rm linux-rpi4.bin
rm linux-rpi4.elf
make  \
    IMAGE=/work/Image \
    DTB=../rpi4-ws/rpi4.dtb \
    TARGET=linux-rpi4.bin \
    CROSS_COMPILE=aarch64-none-elf- \
    ARCH=aarch64
```

> notice that this step has changed, so instead of using the kernel
> from `linux/build-aarch64/arch/arm64/boot/` we use the kernel from
> our yocto deploy directory.

Reviewer: ```suggestion
Notice that this step has changed, so instead of using the kernel
from `linux/build-aarch64/arch/arm64/boot/` we use the kernel from
our Yocto deploy directory.
```

Use impersonal form.

> Now we are almost ready to build the Hypervisor - just one more change

Reviewer: Use impersonal form

Author: @DaniilKl added [here](https://github.com/zarhus/docs/pull/17/commits/cc397fb05208ebf300f413249be7255d75e23759)

needs to be applied:

```git

> root@565810a48049:/work/crosscon/CROSSCON-Hypervisor# git diff src/arch/armv8/aborts.c

Reviewer: Again, you are showing the diff, but there is no description how to apply it.

Author: @DaniilKl here it's the same thing as in [this thread](https://github.com/zarhus/docs/pull/17#discussion_r2024701921).

If i see a reply there, with a suggestion on what needs to be done then i will add it here.

If you read my reply there and agree with the reasoning, then this also can be closed.

Author: @DaniilKl added [here](https://github.com/zarhus/docs/pull/17/commits/b704988a6e0310c086d6c652ba637d5ab11b1c04)

diff --git a/src/arch/armv8/aborts.c b/src/arch/armv8/aborts.c
index a7f5adc..503dc30 100644
--- a/src/arch/armv8/aborts.c
+++ b/src/arch/armv8/aborts.c
@@ -43,6 +43,7 @@ void internal_abort_handler(uint64_t gprs[]) {

 void aborts_data_lower(uint64_t iss, uint64_t far, uint64_t il)
 {
+    printk("magic printk\n");
     if (!(iss & ESR_ISS_DA_ISV_BIT) || (iss & ESR_ISS_DA_FnV_BIT)) {
         ERROR("no information to handle data abort (0x%x)", far);
     }
root@565810a48049:/work/crosscon/CROSSCON-Hypervisor#
```

> this additional `printk` prevents keeps the UART output going, so that

Reviewer: ```suggestion
This additional `printk` prevents keeps the UART output going, so that
```

Prevents or keeps? Link the appropriate CROSSCON issue here.

Author: @DaniilKl fixed [here](https://github.com/zarhus/docs/pull/17/commits/6a6b2bb3f0433e7a56f50ade584264a53d9fa6d8)

the console can be accessed.

Now we can re-run the commands needed to build the Hypervisor:

```bash
CONFIG_REPO=/work/crosscon/rpi4-ws/configs
cd /work/crosscon

make -C CROSSCON-Hypervisor/ \
	PLATFORM=rpi4 \
	CONFIG_BUILTIN=y \
	CONFIG_REPO=$CONFIG_REPO \
	CONFIG=rpi4-single-vTEE \
	OPTIMIZATIONS=0 \
        SDEES="sdSGX sdTZ" \
	CROSS_COMPILE=aarch64-none-elf- \
        clean

make -C CROSSCON-Hypervisor/ \
	PLATFORM=rpi4 \
	CONFIG_BUILTIN=y \
	CONFIG_REPO=$CONFIG_REPO \
	CONFIG=rpi4-single-vTEE \
	OPTIMIZATIONS=0 \
        SDEES="sdSGX sdTZ" \
	CROSS_COMPILE=aarch64-none-elf- \
        -j`nproc`
```

> after executing these commands, the binary will be at this path:

Reviewer: ```suggestion
After executing these commands, the binary will be at the following path:
```

```bash
/work/crosscon/CROSSCON-Hypervisor/bin/rpi4/builtin-configs/rpi4-single-vTEE/crossconhyp.bin
```

> The last step is to copy it over to host, then onto the SD card:

Reviewer: What `host`? You do not mention that you are inside container.

Reviewer: No, it is not clear. You could add at least `root@565810a48049:/work/crosscon/rpi4-ws#` or smth similar before the commands, to show where you are running the commands.

Author: @DaniilKl added clear logs [here](https://github.com/zarhus/docs/pull/17/commits/6700f08343f5deb7fe26557b7c19fb5e50e28531)

Author: @DaniilKl its more clear now:
https://github.com/zarhus/docs/pull/17/files#diff-adebefe8a0ae122e45c2a88a5175f9d36ca39d30cd00d56d9a7603cc672d3badR229-R243
i put this link because i can't find the exact commit

Reviewer: You should add `root@565810a48049:/work/crosscon/rpi4-ws#` to all commands, that are being executed inside the container.

Author: @DaniilKl this was added and can be visible [here](https://github.com/zarhus/docs/pull/17/files#diff-adebefe8a0ae122e45c2a88a5175f9d36ca39d30cd00d56d9a7603cc672d3badR233-R236), so this thread can be closed

```bash
docker cp crosscon_hv_container:/work/crosscon/CROSSCON-Hypervisor/bin/rpi4/builtin-configs/rpi4-single-vTEE/crossconhyp.bin .
cp crossconhyp.bin /run/media/$USER/boot/ # or wherever the SD card is mounted
```

> Finally, the setup for connections and booting is the same as in the

Reviewer: ```suggestion
The final setup for connections and booting is the same as in the
```

# Zarhus OS on CROSSCON Hypervisor on RPi 4

This guide provides info on how to boot Zarhus OS

> on a Raspberry Pi 4 device running the CROSSCON Hypervisor.

Reviewer 2: Link to crosscon hypervisor would be appreciated.

Author: @pietrushnic added [here](https://github.com/zarhus/docs/pull/17/commits/e6d846ff8848a549cb13153a76f19b718af5edef)

Reviewer: The link has been added, resolving this thread.

# Zarhus OS on CROSSCON Hypervisor on RPi 4

This guide provides info on how to boot Zarhus OS on a Raspberry Pi 4 device
running the
[CROSSCON Hypervisor](https://github.com/crosscon/CROSSCON-Hypervisor).

## Build CROSSCON Hypervisor

It is recommended first to follow the
[default demo](https://github.com/crosscon/CROSSCON-Hypervisor-and-TEE-Isolation-Demos/blob/67eb25c97dda2457ca08fa1a10ea68c9c6095b05/rpi4-ws/README.md)
using the
[docker environment](https://github.com/crosscon/CROSSCON-Hypervisor-and-TEE-Isolation-Demos/blob/b437f7ecfe2a24bed6219e805ad0133882b832f4/env/README.md)
to isolate the dependencies.

Make sure to save the content of the prepared boot partition, so that

> it will not be lost, as it will be needed later.

Author: @DaniilKl added [here](https://github.com/crosscon/CROSSCON-Hypervisor-and-TEE-Isolation-Demos/blob/67eb25c97dda2457ca08fa1a10ea68c9c6095b05/rpi4-ws/README.md#setup-board)

32-bit version of the system - `raspberrypi4-64` needs to be used in order to
get a 64-bit version, so that it's compatible with the hypervisor.

## The final image

When the Zarhus build finishes, flash the SD card:

```bash
cd build/tmp/deploy/images/raspberrypi4-64/
sudo bmaptool copy --bmap zarhus-base-image-debug-raspberrypi4-64.rootfs.wic.bmap zarhus-base-image-debug-raspberrypi4-64.rootfs.wic.gz /dev/sd[X]
```

!!! note
    Replace `/dev/sd[X]` with the actual card.

Once that finishes, reinsert the card and delete everything from the `/boot` partition,
and replace it with the content produced by the CROSSCON Hypervisor demo step
above.

The kernel also needs to be swapped for the Zarhus OS kernel. It can be done by
first copying the kernel to the container (make sure to copy the actual kernel,
not the symlink):

```bash
cd build/tmp/deploy/images/raspberrypi4-64/
file Image # This will show the actual file the "Image" symlink points to.
# Use that filename in the next command to ensure you're copying the real
# kernel file, not the symlink
docker cp Image-1-6.6.22+git0+6a24861d65_c04af98514-r0-raspberrypi4-64-20250226164115.bin crosscon_hv_container:/work/Image
```

> We will have to recompile the Hypervisor, starting from

Author: @DaniilKl added [here](https://github.com/zarhus/docs/pull/17/commits/49d5c3570f32842d608a1ca58fef493d7ca7cf55)

+        bootargs = "8250.nr_uarts=8 root=/dev/mmcblk1p2 rw rootwait console=ttyS1,115200 earlycon clk_ignore_unused ip=192.168.42.15 carrier_timeout=0";
     };

     reserved-memory {
@@ -824,7 +824,7 @@
             pinctrl-names = "default";
             pinctrl-0 = <0x07 0x08>;
             uart-has-rtscts;
-            status = "okay";
+            status = "disabled";

             bluetooth {
                 compatible = "brcm,bcm43438-bt";
@@ -1402,14 +1402,14 @@
         #address-cells = <0x02>;
         #size-cells = <0x01>;
         ranges = <0x00 0x7e000000 0x00 0xfe000000 0x1800000>;
-        dma-ranges = <0x00 0xc0000000 0x00 0x00 0x40000000>;
+        dma-ranges = <0x0 0x0 0x0 0x0 0xfc000000>;

         emmc2@7e340000 {
             compatible = "brcm,bcm2711-emmc2";
             reg = <0x00 0x7e340000 0x100>;
             interrupts = <0x00 0x7e 0x04>;
             clocks = <0x06 0x33>;
-            status = "disabled";
+            status = "okay";
             vqmmc-supply = <0x1e>;
             vmmc-supply = <0x1f>;
             broken-cd;
root@565810a48049:/work/crosscon/rpi4-ws#
```

Then run the following commands:

```bash
cd /work/crosscon

> dtc -I dts -O dtb rpi4-ws/rpi4.dts > rpi4-ws/rpi4.dtb

Reviewer: Add `cd lloader` after `dtc`.

Author: @DaniilKl added [here](https://github.com/zarhus/docs/pull/17/commits/d07e92c65caeabd0ab7ff209effa2cad90270402)

    TARGET=linux-rpi4.bin \
    CROSS_COMPILE=aarch64-none-elf- \
    ARCH=aarch64
```

Notice that this step has changed, so instead of using the kernel
from `linux/build-aarch64/arch/arm64/boot/` we use the kernel from
our Yocto deploy directory.

Now the Hypervisor is almost ready to be built - just one more change
needs to be applied:

```git
root@565810a48049:/work/crosscon/CROSSCON-Hypervisor# git diff src/arch/armv8/aborts.c
diff --git a/src/arch/armv8/aborts.c b/src/arch/armv8/aborts.c
index a7f5adc..503dc30 100644
--- a/src/arch/armv8/aborts.c
+++ b/src/arch/armv8/aborts.c
@@ -43,6 +43,7 @@ void internal_abort_handler(uint64_t gprs[]) {

 void aborts_data_lower(uint64_t iss, uint64_t far, uint64_t il)
 {
+    printk("magic printk\n");
     if (!(iss & ESR_ISS_DA_ISV_BIT) || (iss & ESR_ISS_DA_FnV_BIT)) {
         ERROR("no information to handle data abort (0x%x)", far);
     }
root@565810a48049:/work/crosscon/CROSSCON-Hypervisor#
```

This additional `printk` keeps the UART output going, so that the console can
be accessed. Right now, according to
[this issue](https://github.com/crosscon/CROSSCON-Hypervisor-and-TEE-Isolation-Demos/issues/8#issuecomment-2702293550)

> this is the only fix.

Author: @DaniilKl added [here](https://github.com/zarhus/docs/pull/17/commits/356bfe8a13037e80357362be41c4096e2274dc0a)

Now we can re-run the commands needed to build the Hypervisor:

```bash
CONFIG_REPO=/work/crosscon/rpi4-ws/configs
cd /work/crosscon

make -C CROSSCON-Hypervisor/ \
	PLATFORM=rpi4 \
	CONFIG_BUILTIN=y \
	CONFIG_REPO=$CONFIG_REPO \
	CONFIG=rpi4-single-vTEE \
	OPTIMIZATIONS=0 \
        SDEES="sdSGX sdTZ" \
	CROSS_COMPILE=aarch64-none-elf- \
        clean

make -C CROSSCON-Hypervisor/ \
	PLATFORM=rpi4 \
	CONFIG_BUILTIN=y \
	CONFIG_REPO=$CONFIG_REPO \
	CONFIG=rpi4-single-vTEE \
	OPTIMIZATIONS=0 \
        SDEES="sdSGX sdTZ" \
	CROSS_COMPILE=aarch64-none-elf- \
        -j`nproc`
```

After executing these commands, the binary will be at the following path:

```bash
/work/crosscon/CROSSCON-Hypervisor/bin/rpi4/builtin-configs/rpi4-single-vTEE/crossconhyp.bin
```

The last step is to copy it over to host, then onto the SD card:

```bash
docker cp crosscon_hv_container:/work/crosscon/CROSSCON-Hypervisor/bin/rpi4/builtin-configs/rpi4-single-vTEE/crossconhyp.bin .

> cp crossconhyp.bin /run/media/$USER/boot/ # or wherever the SD card is mounted

Author: @DaniilKl added [here](https://github.com/zarhus/docs/pull/17/commits/ca0ba7df16aa987f1e500f5a543be9267648f84a)

```

The final setup for connections and booting is the same as in the

> [last steps of the Hypervisor demo](https://github.com/crosscon/CROSSCON-Hypervisor-and-TEE-Isolation-Demos/tree/master/rpi4-ws/#setup-board).

Author: @DaniilKl added [here](https://github.com/zarhus/docs/pull/17/commits/492a0246ec8a0097179e22ae44c43f8f3be12e6c)

running the
[CROSSCON Hypervisor](https://github.com/crosscon/CROSSCON-Hypervisor).

## Build CROSSCON Hypervisor

It is recommended first to follow the
[default demo](https://github.com/crosscon/CROSSCON-Hypervisor-and-TEE-Isolation-Demos/blob/67eb25c97dda2457ca08fa1a10ea68c9c6095b05/rpi4-ws/README.md)
using the
[docker environment](https://github.com/crosscon/CROSSCON-Hypervisor-and-TEE-Isolation-Demos/blob/b437f7ecfe2a24bed6219e805ad0133882b832f4/env/README.md)
to isolate the dependencies.
Make sure to save the content of the prepared boot partition, so that
it will not be lost, as it will be needed later.

## Build Zarhus

Zarhus will also have to be built (on the host).
Make sure to run this exact build command:

```bash
SHELL=/bin/bash KAS_MACHINE=raspberrypi4-64 kas-container build meta-zarhus/kas/common.yml:meta-zarhus/kas/cache.yml:meta-zarhus/kas/debug.yml:meta-zarhus/kas/rpi.yml
```

The `KAS_MACHINE` variable is very important here, `raspberrypi4` would build a
32-bit version of the system - `raspberrypi4-64` needs to be used in order to
get a 64-bit version, so that it's compatible with the hypervisor.

> ## The final image

Reviewer: After reading entire document I think this should be changed to:

```suggestion
## Replacing kernel
```

running the
[CROSSCON Hypervisor](https://github.com/crosscon/CROSSCON-Hypervisor).

## Build CROSSCON Hypervisor

It is recommended first to follow the
[default demo](https://github.com/crosscon/CROSSCON-Hypervisor-and-TEE-Isolation-Demos/blob/67eb25c97dda2457ca08fa1a10ea68c9c6095b05/rpi4-ws/README.md)
using the
[docker environment](https://github.com/crosscon/CROSSCON-Hypervisor-and-TEE-Isolation-Demos/blob/b437f7ecfe2a24bed6219e805ad0133882b832f4/env/README.md)
to isolate the dependencies.
Make sure to save the content of the prepared boot partition, so that
it will not be lost, as it will be needed later.

## Build Zarhus

Zarhus will also have to be built (on the host).
Make sure to run this exact build command:

```bash
SHELL=/bin/bash KAS_MACHINE=raspberrypi4-64 kas-container build meta-zarhus/kas/common.yml:meta-zarhus/kas/cache.yml:meta-zarhus/kas/debug.yml:meta-zarhus/kas/rpi.yml
```

!!! note

    The `KAS_MACHINE` variable is very important here, `raspberrypi4` would build a
    32-bit version of the system - `raspberrypi4-64` needs to be used in order to
    get a 64-bit version, so that it's compatible with the hypervisor.

## The final image

> When the Zarhus build finishes, flash the SD card:

Reviewer: ```suggestion
When the Zarhus OS build finishes, flash the SD card:
```

```bash
CONFIG_REPO=/work/crosscon/rpi4-ws/configs
cd /work/crosscon

make -C CROSSCON-Hypervisor/ \
	PLATFORM=rpi4 \
	CONFIG_BUILTIN=y \
	CONFIG_REPO=$CONFIG_REPO \
	CONFIG=rpi4-single-vTEE \
	OPTIMIZATIONS=0 \
        SDEES="sdSGX sdTZ" \
	CROSS_COMPILE=aarch64-none-elf- \
        clean

make -C CROSSCON-Hypervisor/ \
	PLATFORM=rpi4 \
	CONFIG_BUILTIN=y \
	CONFIG_REPO=$CONFIG_REPO \
	CONFIG=rpi4-single-vTEE \
	OPTIMIZATIONS=0 \
        SDEES="sdSGX sdTZ" \
	CROSS_COMPILE=aarch64-none-elf- \
        -j`nproc`
```

After executing these commands, the binary will be at the following path:

```bash
/work/crosscon/CROSSCON-Hypervisor/bin/rpi4/builtin-configs/rpi4-single-vTEE/crossconhyp.bin
```

The second-to-last step is to copy it over to host, then onto the SD card:

```bash
docker cp crosscon_hv_container:/work/crosscon/CROSSCON-Hypervisor/bin/rpi4/builtin-configs/rpi4-single-vTEE/crossconhyp.bin .
cp crossconhyp.bin /run/media/$USER/boot/
```

> Note: the SD card doesnt necessarily have to be mounted at
> `/run/media/$USER/boot`, so make sure to check where it's actually mounted
> by using `lsblk` and checking the location.

> And finally, the SD card has to be reinserted, and the second partition (the
> one containing the rootfs has to be mounted). This is necessary in order to
> edit the `/etc/fstab` file, which tells the kernel where the `/boot` partition
> is.

Reviewer: ```suggestion
And finallymthe second partition (the one containing the rootfs has to be
mounted). This is necessary in order to edit the `/etc/fstab` file, which tells
the kernel where the `/boot` partition is.
```

Once the second partition has been mounted, make sure the `/etc/fstab` file
looks exactly like this (this command assumes that the second partition
has been manually mounted using `mount` at `/mnt` directory):

```bash
user in ~ λ cat /mnt/etc/fstab
# stock fstab - you probably want to override this with a machine specific one

/dev/root            /                    auto       defaults              1  1
proc                 /proc                proc       defaults              0  0
devpts               /dev/pts             devpts     mode=0620,ptmxmode=0666,gid=5      0  0
tmpfs                /run                 tmpfs      mode=0755,nodev,nosuid,strictatime 0  0
tmpfs                /var/volatile        tmpfs      defaults              0  0

# uncomment this if your device has a SD/MMC/Transflash slot
#/dev/mmcblk0p1       /media/card          auto       defaults,sync,noauto  0  0

/dev/mmcblk1p1	/boot	vfat	defaults	0	0
user in ~ λ
```

By default, `/boot` is specified for `/dev/mmcblk0p1`, which works for the

> purpose of using Zarhus without the CROSSCON Hypervisor, but in this case

Reviewer: ```suggestion
purpose of using Zarhus OS without the CROSSCON Hypervisor, but in this case
```

it has to be edited to point to the correct partition, which is
`/dev/mmcblk1p1`

The final setup for connections and booting is the same as in the

> [last steps of the Hypervisor demo](https://github.com/crosscon/CROSSCON-Hypervisor-and-TEE-Isolation-Demos/blob/67eb25c97dda2457ca08fa1a10ea68c9c6095b05/rpi4-ws/README.md#setup-board).

Reviewer: ```suggestion
[last steps of the Hypervisor demo](https://github.com/crosscon/CROSSCON-Hypervisor-and-TEE-Isolation-Demos/blob/67eb25c97dda2457ca08fa1a10ea68c9c6095b05/rpi4-ws/README.md#setup-board).

## Booting
```

And show some logs for reader to compare (do not include the magic printk).

Author: @DaniilKl added [here](https://github.com/zarhus/docs/pull/17/commits/512a542df339a25b42f238cfce63021885673a2b) and [here](https://github.com/zarhus/docs/pull/17/commits/f3463308d8b9cd54a8f4ef191927f8d678cbd35d)

Author: @DaniilKl added [here](https://github.com/zarhus/docs/pull/17/commits/512a542df339a25b42f238cfce63021885673a2b) and [here](https://github.com/zarhus/docs/pull/17/commits/f3463308d8b9cd54a8f4ef191927f8d678cbd35d)

             bluetooth {
                 compatible = "brcm,bcm43438-bt";
@@ -1402,14 +1402,14 @@
         #address-cells = <0x02>;
         #size-cells = <0x01>;
         ranges = <0x00 0x7e000000 0x00 0xfe000000 0x1800000>;
-        dma-ranges = <0x00 0xc0000000 0x00 0x00 0x40000000>;
+        dma-ranges = <0x0 0x0 0x0 0x0 0xfc000000>;

         emmc2@7e340000 {
             compatible = "brcm,bcm2711-emmc2";
             reg = <0x00 0x7e340000 0x100>;
             interrupts = <0x00 0x7e 0x04>;
             clocks = <0x06 0x33>;
-            status = "disabled";
+            status = "okay";
             vqmmc-supply = <0x1e>;
             vmmc-supply = <0x1f>;
             broken-cd;
root@565810a48049:/work/crosscon/rpi4-ws#
```

!!! note

    This `git diff` was put here in order to outline the changes that
    need to be made to the `rpi4.dts` file. It can work as a patch to be applied,
    but if it doesn't apply, installing some sort of text editor on the container
    in order to apply the changes manually will be necessary:
    ```bash
    apt-get update
    apt-get upgrade
    apt-get install vim nano
    ```

The reasoning for these changes is as follows:

> * `8250.nr_uarts=8` tells the `8250` serial driver to allocate up to 8 serial

Reviewer: This does not render right:
![image](https://github.com/user-attachments/assets/e9534b3b-801d-43a9-9f53-ca2fa0c31cc2)

Author: @DaniilKl sorry about that, fixed [here](https://github.com/zarhus/docs/pull/17/commits/755395f5ecaef3af0189efc9a71a5bb5af447a3f)

ports. This is necessary, because if this line is not here, logs from OS level
will not show up on the console.
* `root=/dev/mmcblk1p2` tells the kernel that there is a `/root` partition
at that location. By default, the CROSSCON Hypervisor doesn't have a `rootfs`
or a partition for it, so this has to be added.
* `rw` ensures that the `rootfs` gets mounted in `read-write` mode, rather
than `read-only`.

> * `rootwait` tells the kernel to wait until the root device (in our case,

Reviewer: ```suggestion
* `rootwait` tells the kernel to wait until the root device (in this case,
```

`/dev/mmcblk1p2`) is detected and available before trying to mount it. There
can be a delay when enumerating the storage device, so this stops the kernel
from giving up too early.
* `console=ttyS1,115200` sets the right console, this enables access to the
console and logs from OS level.
* setting `status = "disabled"` for the Bluetooth driver, while technically not
necessary, is a good practice as it can disrupt the serial console setup.
* Finally, setting the `dma-ranges` and `status="okay"` for the `emmc2bus`
enables us to use the whole SD card storage device.

Then run the following commands:

```bash
cd /work/crosscon
dtc -I dts -O dtb rpi4-ws/rpi4.dts > rpi4-ws/rpi4.dtb
cd lloader
rm linux-rpi4.bin
rm linux-rpi4.elf
make  \
    IMAGE=/work/Image \
    DTB=../rpi4-ws/rpi4.dtb \
    TARGET=linux-rpi4.bin \
    CROSS_COMPILE=aarch64-none-elf- \
    ARCH=aarch64
```

Notice that this step has changed, so instead of using the kernel

> from `linux/build-aarch64/arch/arm64/boot/` we use the kernel from

Reviewer: Use impersonal form

Author: @DaniilKl fixed [here](https://github.com/zarhus/docs/pull/17/commits/ac52eae1e591f5bbf48e857a043e3fb8b61f57e8)

# Zarhus OS on CROSSCON Hypervisor on RPi 4

This guide provides info on how to boot Zarhus OS on a Raspberry Pi 4 device
running the
[CROSSCON Hypervisor](https://github.com/crosscon/CROSSCON-Hypervisor).

## Build CROSSCON Hypervisor

First, follow the
[default demo](https://github.com/crosscon/CROSSCON-Hypervisor-and-TEE-Isolation-Demos/blob/67eb25c97dda2457ca08fa1a10ea68c9c6095b05/rpi4-ws/README.md)
using the
[docker environment](https://github.com/crosscon/CROSSCON-Hypervisor-and-TEE-Isolation-Demos/blob/b437f7ecfe2a24bed6219e805ad0133882b832f4/env/README.md)

> to isolate the dependencies.

Reviewer: Line length should be close but do not exceed 80 characters

    apt-get update
    apt-get upgrade
    apt-get install vim nano
    ```

The reasoning for these changes is as follows:

* `8250.nr_uarts=8` tells the `8250` serial driver to allocate up to 8 serial
ports. This is necessary, because if this line is not here, logs from OS level
will not show up on the console.
* `root=/dev/mmcblk1p2` tells the kernel that there is a `/root` partition
at that location. By default, the CROSSCON Hypervisor doesn't have a `rootfs`
or a partition for it, so this has to be added.
* `rw` ensures that the `rootfs` gets mounted in `read-write` mode, rather
than `read-only`.
* `rootwait` tells the kernel to wait until the root device (in our case,
`/dev/mmcblk1p2`) is detected and available before trying to mount it. There
can be a delay when enumerating the storage device, so this stops the kernel
from giving up too early.
* `console=ttyS1,115200` sets the right console, this enables access to the
console and logs from OS level.
* setting `status = "disabled"` for the Bluetooth driver, while technically not
necessary, is a good practice as it can disrupt the serial console setup.
* Finally, setting the `dma-ranges` and `status="okay"` for the `emmc2bus`

> enables us to use the whole SD card storage device.

Reviewer: Use impersonal form

Author: @DaniilKl fixed [here](https://github.com/zarhus/docs/pull/17/commits/ac52eae1e591f5bbf48e857a043e3fb8b61f57e8)

our Yocto deploy directory.

Now the CROSSCON Hypervisor is almost ready to be built - just one more change
needs to be applied:

```git
root@565810a48049:/work/crosscon/CROSSCON-Hypervisor# git diff src/arch/armv8/aborts.c
diff --git a/src/arch/armv8/aborts.c b/src/arch/armv8/aborts.c
index a7f5adc..503dc30 100644
--- a/src/arch/armv8/aborts.c
+++ b/src/arch/armv8/aborts.c
@@ -43,6 +43,7 @@ void internal_abort_handler(uint64_t gprs[]) {

 void aborts_data_lower(uint64_t iss, uint64_t far, uint64_t il)
 {
+    printk("magic printk\n");
     if (!(iss & ESR_ISS_DA_ISV_BIT) || (iss & ESR_ISS_DA_FnV_BIT)) {
         ERROR("no information to handle data abort (0x%x)", far);
     }
root@565810a48049:/work/crosscon/CROSSCON-Hypervisor#
```

!!! note

    Again, just like in the above `git diff`, a text editor might be
    necessary to use in order to apply this change.

This additional `printk` keeps the UART output going, so that the console can
be accessed. Right now, according to
[this issue](https://github.com/crosscon/CROSSCON-Hypervisor-and-TEE-Isolation-Demos/issues/8#issuecomment-2702293550)
this is the only workaround.

> Now we can re-run the commands needed to build the Hypervisor:

Reviewer: Use impersonal form

Author: @DaniilKl fixed [here](https://github.com/zarhus/docs/pull/17/commits/ac52eae1e591f5bbf48e857a043e3fb8b61f57e8)
